[skip-ci] Packit: New workflow for downstream Fedora and CentOS Stream 10

lsm5 opened the pull request:
This commit enables:
- upstream copr build jobs on PRs
- rpm builds on podman-next copr after every commit to main
- Fedora and CentOS Stream 10 downstream update jobs on every upstream
  release
- Fetch RPM-GPG-KEY-redhat-release from https://access.redhat.com/security/data/fd431d51.txt
- Config file patching during rpm build is managed via `rpm/update-config-files.sh`,
  modified from the original script at
  https://gitlab.com/redhat/centos-stream/rpms/containers-common/-/blob/c9s/update.sh
  and modified to include Fedora and RHEL-10


<!--- Please read the [contributing guidelines](https://github.com/containers/common-files/blob/master/.github/CONTRIBUTING.md) before proceeding --->

lsm5 commented:
@rhatdan @Luap99 @cevich @edsantiago @jnovy @mheon @mtrmac @TomSweeneyRedHat PTAL

podman and toolbox work fine for me with the rpms generated in the rawhide copr build job. Would be great if you could give these a try too.

You can try this out with:
```
sudo dnf -y copr enable packit/containers-common-1960
sudo dnf -y update containers-common-extra
```

lsm5 commented:
cc @inknos if you'd like to try this out too

lsm5 commented on rpm/containers-common.spec:
```diff
@@ -0,0 +1,222 @@
+%endif
+# DO NOT TOUCH the Version string!
+# The TRUE source of this specfile is:
+# https://github.com/containers/podman/blob/main/rpm/podman.spec
```
I need to `s/podman/containers-common/` here. Will fix after I have some reviews

Luap99 commented on .github/workflows/validate.yml:
```diff
@@ -22,7 +22,8 @@ jobs:
       run: pip install codespell
     - name: run codespell
-      run: codespell --dictionary=-
+      # passt is a dependency listed in rpm/containers-common.spec
+      run: codespell --dictionary=- -L passt
```
just add the exception to `.codespellrc`

Luap99 commented on rpm/containers-common.spec:
```diff
@@ -0,0 +1,222 @@
+# Buildah/Podman/Skopeo.
+
+# Replace `main` below with upstream version tags before cutting a new release of c/common
+%global image_branch main
+%global storage_branch main
+%global shortnames_branch main
```
Who is going to do that? I think we have the issue today that many config files do not actually match the tagged versions 1 to 1.

go.mod should already specify the correct version so can we use this to automate this somehow?

lsm5 replied:
added a packit action that will run `rpm/update-lib-versions.sh` on downstream PRs. We can only test this once downstream PRs are created though.

Luap99 commented on rpm/containers-common.spec:
```diff
@@ -0,0 +1,222 @@
+cp %{SOURCE13} .
+cp %{SOURCE14} .
+
+# Patch storage.conf
+sed -i -e 's/^driver.*=.*/driver = "overlay"/' -e 's/^mountopt.*=.*/mountopt = "nodev,metacopy=on"/' \
+        -e '/additionalimage.*/a "/usr/lib/containers/storage",' \
+%if %{defined copr_build} || (%{defined fedora} && 0%{?fedora} > 40)
+        -e 's/^pull_options.*=.*/pull_options = {enable_partial_images = \"true\", use_hard_links = \"false\", ostree_repos="", convert_images = "false"}/' \
+        -e 's/# use_composefs.*/use_composefs = "false"/g' \
+%endif
+        storage.conf
+
+# Patch seccomp.json
+[ `grep "keyctl" pkg/seccomp/seccomp.json | wc -l` == 0 ] && sed -i '/\"kill\",/i \
+                                "keyctl",' pkg/seccomp/seccomp.json
+sed -i '/\"socketcall\",/i \
+                                "socket",' pkg/seccomp/seccomp.json
+
+# Patch registries.conf
+sed -i 's/^#.*unqualified-search-registries.*=.*/unqualified-search-registries = ["registry.fedoraproject.org", "registry.access.redhat.com", "docker.io"]/g' \
+        registries.conf
+grep '^short-name-mode="enforcing"' registries.conf && echo -e '\nshort-name-mode="enforcing"' >> registries.conf
+
+# Patch containers.conf
+sed -i -e 's/^#.*log_driver.*=.*/log_driver = "journald"/' \
+    -e 's/^#.*compression_format.*=.*/compression_format = "zstd:chunked"/' \
+    pkg/config/containers.conf
```
I think we should review this? Why do we patch our own upstream config files so heavily.
I understand that some of them make sense, i.e. adding fedora registries but other are not so obvious. I know this is juts copying the function out of the rpm of today but I really think these should have a clear explanation WHY they are patched in this way.

lsm5 replied:
@rhatdan I lifted these from fedora dist-git. How many of these patches are safe to merge/enable upstream?

inknos replied:
my only 2 cents on this PR would be that scripts could be handled out of the specfile. that's mostly a personal preference as I don't think there's any recommendation on this.

lsm5 replied:
@inknos ack thanks. We've been doing something similar in dist-git so far https://src.fedoraproject.org/rpms/containers-common/blob/rawhide/f/update.sh . Let's check if we can reduce these further.

mtrmac replied:
(I know, this is not the topic of the PR, feel free to ignore.)

RPM has a native %patch (%autopatch?) mechanism. It seems to me that doing this with manual `sed` all much less readable, and more risky to do something unexpected without us noticing, and I don’t see any benefit.

I tend to think that patch conflicts and application failures are actually our friends.

Does anyone here immediately know what `-e '/additionalimage.*/a "/usr/lib/containers/storage",'` does without referring to the `storage.conf` being edited?!

---

I also agree that if Fedora downstream is using non-default options, that is quite possible but perhaps non-obvious. So having some documentation about which changes are
- forward-ports where we force rawhide to be more experimental than upstream releases
- backports where upstream is ahead of a stable version in a RPM
- long-term differences
along with reasons, would help future maintainers. Or, well, _current_ maintainers asking “How many of these patches are safe to merge/enable upstream?”.

Especially if we move the location where this is maintained, thus requiring extra knowledge to find the original `git` history.

---

FWIW, non-authoritative, just to start the discussion:
- ` -e '/additionalimage.*/a "/usr/lib/containers/storage",' \` feels like CoreOS/image-based-OS specific, not a good upstream default (but I’m not 100% sure)
- `enable_partial_images` etc., `"zstd:chunked"` are experimental and _not_ upstream-ready at this point (ref . https://github.com/containers/storage/issues/1888 )
- `unqualified-search-registries` are Fedora-specific to keep compatibility with older Fedora users, less secure than upstream default
- I don’t immediately know anything about the others (and I didn’t go looking)

lsm5 replied:
> Does anyone here immediately know what `-e '/additionalimage.*/a "/usr/lib/containers/storage",'` does without referring to the `storage.conf` being edited?!

there's storage.conf on line 115. Sorry if the readability is poor currently.

And ack other suggestions too.

inknos replied:
> RPM has a native %patch (%autopatch?) mechanism. It seems to me that doing this with manual sed is all much less readable, and more risky to do something unexpected without us noticing, and I don’t see any benefit.

Sorry for being late to the discussion. +1 for migrating to spec file patches as much as possible in the future.

lsm5 commented on contrib/redhat/RPM-GPG-KEY-redhat-release:
```diff
@@ -0,0 +1,34 @@
+pub   4096R/FD431D51 2009-10-22
```
I think it'd be nice to fetch the key from the authoritative source instead of storing a copy here.

TomSweeneyRedHat commented:
It LGTM, but way out of my wheelhouse.  I'd like other eyeballs before merge.

lsm5 commented:
> @lsm5 FYI in c9s/RHEL we use a dedicated script with a nice wrapper to configure things:
> 
> https://gitlab.com/redhat/centos-stream/rpms/containers-common/-/blob/c9s/update.sh?ref_type=heads
> 
> Maybe something like this would make configurations more readable?

@jnovy @rhatdan how much of this could be upstreamed as well ?

mtrmac commented on rpm/containers-common.spec:
```diff
@@ -0,0 +1,228 @@
+# Fetch RPM-GPG-KEY-redhat-release from authoritative source instead of storing
+# a copy in repo or dist-git. Depending on distribution-gpg-keys rpm is also
+# not an option because that package doesn't exist on CentOS Stream.
+Source15: %{github_raw}/xsuchy/distribution-gpg-keys/main/keys/redhat/RPM-GPG-KEY-redhat9-release
```
He’s a friend, but not particularly the authoritative source. Is it available from https://access.redhat.com/security/team/key ?

lsm5 replied:
lol :D . That's the repo being used by the `distribution-gpg-keys` fedora package and what was pointed to me in the rhel-centos internal channel.

RE: the link you sent, https://access.redhat.com/security/data/fd431d51.txt seems to be available publicly, I could use that one, unless there's a floating link to always point to the current one. 

Thanks a lot!

mtrmac replied:
I’m sorry, I had no idea Fedora already trusts that repo.

(The actual builds can’t fetch from the internet anyway, can they?)

lsm5 replied:
switched to access.redhat.com link. Thanks again @mtrmac

lsm5 replied:
> (The actual builds can’t fetch from the internet anyway, can they?)

They should be able to fetch if it's listed as a SourceN, but good point, best to verify.

rhatdan commented on rpm/containers-common.spec:
```diff
@@ -0,0 +1,227 @@
+# Patch storage.conf
+sed -i -e 's/^driver.*=.*/driver = "overlay"/' -e 's/^mountopt.*=.*/mountopt = "nodev,metacopy=on"/' \
+        -e '/additionalimage.*/a "/usr/lib/containers/storage",' \
+%if %{defined copr_build} || (%{defined fedora} && 0%{?fedora} > 40)
```
What about RHEL 10?

jnovy commented:
> > @lsm5 FYI in c9s/RHEL we use a dedicated script with a nice wrapper to configure things:
> > https://gitlab.com/redhat/centos-stream/rpms/containers-common/-/blob/c9s/update.sh?ref_type=heads
> > Maybe something like this would make configurations more readable?
> 
> @jnovy @rhatdan how much of this could be upstreamed as well ?

Assuming all of it - if we have a reliable way who to detect RHELX/FedoraX, etc from the script.

edsantiago commented on rpm/containers-common.spec:
```diff
@@ -0,0 +1,227 @@
+cp %{SOURCE14} .
+cp %{SOURCE15} RPM-GPG-KEY-redhat-release
+
+# Patch storage.conf
```
Could you explain _why_ these patches (this and all the ones below) are necessary?

lsm5 replied:
added some comments to `rpm/update-config-files.sh` PTAL.

lsm5 commented:
I'll open individual PRs to upstream changes for storage.conf, registries.conf and containers.conf for easier reviewing. If we still need to patch here, I can include refs to those reviews in this PR.

mtrmac commented:
Meta-comment: Do we want to block this PR on the spec restructuring? There are certainly things in there (I ignorantly think) I’d prefer changing, but IIRC we rather wanted to land the Packit integration as a feature valuable in the short-term.

Sure, one possible answer is “a few more days doesn’t matter either way”.

lsm5 commented:
> Meta-comment: Do we want to block this PR on the spec restructuring? There are certainly things in there (I ignorantly think) I’d prefer changing, but IIRC we rather wanted to land the Packit integration as a feature valuable in the short-term.
> 
> Sure, one possible answer is “a few more days doesn’t matter either way”.

@TomSweeneyRedHat @mheon when do we want to cut a new c/common for all the vendor dancing stuff?

mheon commented:
If we need one, late this week.

On Mon, Apr 29, 2024 at 13:57 Lokesh Mandvekar ***@***.***>
wrote:

> Meta-comment: Do we want to block this PR on the spec restructuring? There
> are certainly things in there (I ignorantly think) I’d prefer changing, but
> IIRC we rather wanted to land the Packit integration as a feature valuable
> in the short-term.
>
> Sure, one possible answer is “a few more days doesn’t matter either way”.
>
> @TomSweeneyRedHat <https://github.com/TomSweeneyRedHat> @mheon
> <https://github.com/mheon> when do we want to cut a new c/common for all
> the vendor dancing stuff?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/containers/common/pull/1960#issuecomment-2083325042>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB3AOCAGFSWKAHL5KHKNYS3Y72C2DAVCNFSM6AAAAABGNOPLSWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAOBTGMZDKMBUGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

lsm5 commented:
Good for further review. PTAL.

For testing:
```
sudo dnf -y copr enable packit/containers-common-1960
sudo dnf -y update containers-common containers-common-extra
```
I moved all the patches to `rpm/update-config-files.sh`, modified from the original at https://gitlab.com/redhat/centos-stream/rpms/containers-common/-/blob/c9s/update.sh by @jnovy .  The rpm spec will run this script in the `%prep` stage.

Things I'm hoping we can keep for later:
- Reduce patching
- `%patch` v/s `script` for patch management

edsantiago commented on rpm/update-config-files.sh:
```diff
@@ -0,0 +1,61 @@
+
+ensure pkg/config/containers.conf   runtime             \"crun\"
+ensure pkg/config/containers.conf   log_driver          \"journald\"
+ensure pkg/config/containers.conf   compression_format  \"zstd:chunked\"         
```
Eek! trailing whitespace! I'm alarmed that this repo does not have a block against that.

edsantiago commented on rpm/update-config-files.sh:
```diff
@@ -0,0 +1,61 @@
+
+
+# Enable seccomp support keyctl and socketcall
+[ `grep \"keyctl\", pkg/seccomp/seccomp.json | wc -l` == 0 ] && sed -i '/\"kill\",/i \
```
Very hard to grok; see my diff for an alternative

edsantiago commented on rpm/update-config-files.sh:
```diff
@@ -0,0 +1,61 @@
+[ `grep \"socket\", pkg/seccomp/seccomp.json | wc -l` == 0 ] && sed -i '/\"socketcall\",/i \
+				"socket",' pkg/seccomp/seccomp.json
+
+FEDORA=$(rpm --eval %{?fedora})
+RHEL=$(rpm --eval %{?rhel})
+COPR=$(rpm --eval %{?copr_username})
```
Suggest using single quotes, to eliminate the risk of the shell hijacking the curly braces (and barfing)

edsantiago commented on rpm/update-config-files.sh:
```diff
@@ -0,0 +1,61 @@
+COPR=$(rpm --eval %{?copr_username})
+
+# Set search registries
+if $FEDORA; then
```
This does not work: in non-fedora, you'd get `if ` (if blank), syntax error, barf. See diff.

edsantiago commented on rpm/containers-common.spec:
```diff
@@ -0,0 +1,206 @@
+cp %{SOURCE15} RPM-GPG-KEY-redhat-release
+
+# Patch config files
+sh rpm/update-config-files.sh
```
bash, not sh

edsantiago reviewed (changes requested):
Some suggestions for readability, and a few actual errors. Try applying this diff
[update-config-files.diff.txt](https://github.com/containers/common/files/15169630/update-config-files.diff.txt)

lsm5 commented:
> Some suggestions for readability, and a few actual errors. Try applying this diff [update-config-files.diff.txt](https://github.com/containers/common/files/15169630/update-config-files.diff.txt)

updated. Thanks a lot for the changes @edsantiago . PTAL

edsantiago commented on rpm/containers-common.spec:
```diff
@@ -0,0 +1,206 @@
+cp %{SOURCE10} 000-shortnames.conf
+cp %{SOURCE11} .
+cp %{SOURCE12} policy.json
+cp %{SOURCE13} .
```
source13 is `containers-storage.conf.5.md`, should this line be `cp ... docs/.` like all the other `.md` files?

I'm going to be submitting one-by-one review comments because this is a lot to digest. My apologies.

edsantiago replied:
Actually after some thought ... the whole "cp some files to docs" thing is really squicky. Impossible to maintain. Would it make sense to revert back to the old `ls *.5.md` form (line 117)?

lsm5 replied:
> source13 is `containers-storage.conf.5.md`, should this line be `cp ... docs/.` like all the other `.md` files?

ohh thanks. I'll check and update.

> I'm going to be submitting one-by-one review comments because this is a lot to digest. My apologies.

Cool with me :) . Appreciate the reviews.

lsm5 replied:
PTAL. I added comments labelling the files along with each `cp %{SOURCEN}` line. Hope that makes things more readable. Also, sent SOURCE13 to docs/.

edsantiago commented on rpm/containers-common.spec:
```diff
@@ -0,0 +1,206 @@
+cp %{SOURCE14} .
+cp %{SOURCE15} RPM-GPG-KEY-redhat-release
+
+# Patch config files
```
This doesn't explain why. A maintainer would see this and feel forced to jump to that file to figure it out. What would be a comment that could help a reader instantly decide "I need to look deep here" vs "nah, this isn't the cause of my problem"? Maybe something like this? (You can probably think of something much better)
```bash
# Fine-grain distro- and release-specific tuning of config files,
# e.g., seccomp, composefs, registries on different RHEL/Fedora versions
```

lsm5 replied:
Done

edsantiago commented on rpm/containers-common.spec:
```diff
@@ -0,0 +1,237 @@
+Provides: skopeo-containers = %{epoch}:%{version}-%{release}
+Requires: (container-selinux >= 2:2.162.1 if selinux-policy)
+Suggests: fuse-overlayfs
+Source0: %{git0}/archive/v%{version_no_tilde}.tar.gz
```
Have you verified that `version_no_tilde` works as intended? I mean, it _has_ to, otherwise I don't think any of this would work... but it seems to be defined in `/usr/lib/rpm/macros.d/macros.rust-srpm` and I don't know enough about RPM macro processing to know if that file is in the standard include path, or if it exists in RHEL/CentOS, or lots of questions. (Same macro is used in podman, so yeah this really seems like it works, but I want to make triple-sure)

lsm5 replied:
it's already there in podman's spec file and has been working well for podman RCs and stable releases. I included `version_no_tilde` to account for any possible RCs we may wanna do in the future.

lsm5 replied:
Besides, centos 10 packit workflow still needs a manual `centpkg build` after dist-git PR merge, so this is a pretty safe change.

jnovy commented on rpm/containers-common.spec:
```diff
@@ -0,0 +1,237 @@
+cp %{SOURCE14} .
+
+# RPM-GPG-KEY-redhat-release
+cp %{SOURCE15} RPM-GPG-KEY-redhat-release
```
@lsm5 would you mind removing all the cp statements above to avoid duplicate copying in the similar ways as it's done in https://gitlab.com/redhat/centos-stream/rpms/containers-common/-/blob/stream-container-tools-latest-rhel-9.0.0-beta/containers-common.spec?ref_type=heads#L77 ? What I mean is to leave the %prep section empty and do everything via 'install' in %install section to avoid double copying?

lsm5 replied:
will do. Your suggestion looks much better, thanks.

lsm5 replied:
done

jnovy commented on rpm/update-config-files.sh:
```diff
@@ -0,0 +1,61 @@
+
+# Set these on Fedora Rawhide (41+), RHEL 10+, and on all COPR builds
+# regardless of distro
+if [[ -n "$COPR" ]] || [[ "$FEDORA" -gt 40 ]] || [[ "$RHEL" -ge 10 ]]; then
```
Assuming we are going to leave containers-common in c8s/c9s as-is so this part is only relevant to c10s @lsm5 ?

lsm5 replied:
I think (hope) we can forget about c8s given it's going EOL end of this month.

RE: c9s, I had only included RHEL10 because I didn't see those changes in c9s. I guess the `use_composefs=false` is safe for c9s and I'll defer to @rhatdan about `pull_options` .

jnovy commented:
LGTM

edsantiago commented on rpm/containers-common.spec:
```diff
@@ -0,0 +1,200 @@
+install -Dp -m0644 contrib/redhat/registry.access.redhat.com.yaml -t %{buildroot}%{_sysconfdir}/containers/registries.d
+install -Dp -m0644 contrib/redhat/registry.redhat.io.yaml -t %{buildroot}%{_sysconfdir}/containers/registries.d
+
+# install manpages
```
I like the removal of all the `cp` commands, but the replacement is really hairy: super-duper-duplicationy (duplicating filename and the full path). It also skips using `install -m0644`, hence relies on umask. What would you think of
```console
mkdir -p man5
for i in *.5.md; do
    go-md2man -in $i -out man5/$(basename $i .md)
done

....

for i in man5/*.5; do
    install -Dp -m0644 $i -t %{buildroot}%{_mandir}/man5
done
ln -s containerignore.5 %{buildroot}%{_mandir}/man5/.containerignore.5
```

lsm5 replied:
> ```
> mkdir -p man5
> for i in *.5.md; do
>     go-md2man -in $i -out man5/$(basename $i .md)
> done
> ```

This will need the `cp` statements added back for all those manpages from other repos listed as SourceN files, otherwise it's going to complain those files aren't found.

One thing I could do would be to move all the `go-md2man` steps to `%build` stage and then run the install loop in `%install`.
>```
> for i in man5/*.5; do
>    install -Dp -m0644 $i -t %{buildroot}%{_mandir}/man5
> done
> ln -s containerignore.5 %{buildroot}%{_mandir}/man5/.containerignore.5
>```

lsm5 replied:
> > ```
> > mkdir -p man5
> > for i in *.5.md; do
> >     go-md2man -in $i -out man5/$(basename $i .md)
> > done
> > ```
> 
> This will need the `cp` statements added back for all those manpages from other repos listed as SourceN files, otherwise it's going to complain those files aren't found.
> 

Trying something in a new commit. Let's see how that goes.

lsm5 replied:
@edsantiago You're gonna end up with some or the other `cp` or `mv` somewhere to operate on the SourceN files. So, let me know how you'd like to proceed. I could cp / mv the SourceN to docs, build all docs in `%build`, and then install them in `%install` if that seems ok.

edsantiago replied:
I like this version much, much better: it's less duplicationy. With your move of `constainers-storage.conf.5.md` to be contiguous to all the other `.md` files, this is less fragile. LGTM.

edsantiago commented on rpm/containers-common.spec:
```diff
@@ -0,0 +1,204 @@
+mv %{SOURCE9} docs/.
+
+mkdir -p man5
+for i in $(ls docs/*.5.md); do
```
tiny nit: the `$(ls...)` will confuse readers. Much cleaner to just say `for i in docs/*.5.md; do`

lsm5 commented:
@edsantiago PTAL, comments addressed.

@mheon @TomSweeneyRedHat before we cut a new c/common, I'm hoping to have [this change](https://src.fedoraproject.org/rpms/containers-common/pull-request/6#request_diff) from @ktdreyer reviewed, should come in a new PR once this PR is merged.

edsantiago commented:
/lgtm

Thank you for your patience!

lsm5 commented:
@edsantiago thanks, looks like it needs a separate slash-approve too

Luap99 commented:
/approve

lsm5 commented:
/cherrypick v0.58
